Exploration of the Vehicles dataset based on #2

SaumyaSinghal opened the pull request:
Here I have made an analysis of the vehicles.csv dataset. My exploration can be summarised as follows:
* Load the dataset
* Basic exploration and visualisation of the data
* Train and test on an ensemble of classifiers (trees, SVMs, gaussians, discriminant analysis, etc.) using **cross-validation**
* Tables and plots of the accuracies and predictions
* Voting classifier using all the classifiers individually used till now
* Hyper-parameter tuning using grid search
* Train and test on individual models as well as the voting model with tuned hyper-parameters
* Basic comparison between results.

The maximum accuracy I got was ~84% using Quadratic Discriminant Analysis.

Although this is aimed at #2, I think this touches some of the other issues too. I’ll open separate PRs for them soon.

I will keep improving and documenting but I’d be very happy to get a review of my approach :)
I have also tried to modularise and re-use as much code as possible but I’d highly appreciate any suggestion towards that (and also regarding any python tricks to shorten my code, I’m sure I have missed out on a few :D).

dzeber commented on dev/Saumya/algorithm_helpers.py:
```diff
@@ -0,0 +1,321 @@
+    #Removed models without attribute 'predict_proba' (required for vote classifier) and models with a 1.0 correlation to another model
+    #Using default hyper-parameters
+    vote_est = [
+        #Ensemble Methods
```
Maybe turn `get_algorithms` output into a dict, and just list the keys here, so as to avoid reinitializing the models. Not really a performance issues, more code consistency and maintenance.

SaumyaSinghal replied:
That's right, the code looks a lot better now. Thank you :)

dzeber commented on dev/Saumya/algorithm_helpers.py:
```diff
@@ -0,0 +1,321 @@
+
+
+    start_total = time.perf_counter()
+    for clf, param in zip (vote_est, grid_param):
```
I would make `grid_param` into a dict using the same keys as `vote_est` to make it explicit, rather than relying on zip. This will make code updates easier.

SaumyaSinghal replied:
Yes I later realized that even if the lists are inconsistent *zip* would do its work silently, results would be wrong and I would not even know where the error lies. Thank you for pointing this out!

SaumyaSinghal commented:
@dzeber thank you so much for such a detailed review. I was pondering upon this for the last couple of days and tried addressing most of the issues you have pointed out along with working on other aspects of it.

I thought at first that I'll push individual changes and reply to the individual reviews but it seems like a better idea would be to get everything implemented by Tuesday and make all the comments thereafter so that you can just take a single look once you get a chance :)

mlopatka commented:
@SaumyaSinghal please let us know if you are still actively working on this PR. 
I'll add a possible inactive tag for now, if you flag me here we'll review and remove the flag.

SaumyaSinghal commented:
@mlopatka I am really sorry for the delay and inactivity. Please remove the *inactive* flag. We were suddenly told to evacuate our college hostels in wake of COVID-19 and that caused a lot of hassle, shifting and procuring documents throughout last week.
Things have settled now, I am at home and have been able to resume from yesterday. I'll try to finish this as fast as possible. I am parallelly working on the other issues too and I hope to finish some of them soon.

mlopatka commented:
@SaumyaSinghal No worries at all! I hope you are keeping healthy and safe.
I'm using the inactive tags as a way to help @dzeber and I spend our time most effectively. It has been removed from this PR and I'll loop back for a review as soon as I can.

SaumyaSinghal commented:
@mlopatka thank you so much. This cheered me up :D

SaumyaSinghal commented on dev/Saumya/algorithm_helpers.py:
```diff
@@ -0,0 +1,321 @@
+        #set name and parameters
+        MLA_name = alg.__class__.__name__
+        MLA_compare.loc[row_index, 'MLA Name'] = MLA_name
+        MLA_compare.loc[row_index, 'MLA Parameters'] = str(alg.get_params())
```
This was great advice! Once adjusted to this, I am doing this in all my notebooks. It is super easy to append rows (just adding the lists) while keeping the code maintainable :)

SaumyaSinghal commented on dev/Saumya/algorithm_helpers.py:
```diff
@@ -0,0 +1,321 @@
+
+        #save MLA predictions that could be used later
+        alg.fit(X, y)
+        MLA_predict[MLA_name] = alg.predict(X)
```
Yes this was a mistake on part. This actually caused my correlation heatmap to give faulty correlation values from predictions and I could not understand why it was happening. Thank you, it is fixed now.

mlopatka commented:
@SaumyaSinghal please resolve the merge conflicts.
There is a wealth of advice on how to do this in the gitter channel as several of your fellow outreachy applicants have come across this kind of issue.
